framework/src/base/Assembly.C

```diff
@@ -697,6 +701,31 @@ Assembly::setNeighborQRule(QBase * qrule, unsigned int dim)
     it.second->attach_quadrature_rule(qrule);
 }
 
+void
+Assembly::setMortarQRule(Order order)
+{
+  if (order != _qrule_msm->get_order())
+  {
+    // If custom mortar qrule has not yet been specified
+    if (!_custom_mortar_qrule)
+    {
+      _custom_mortar_qrule = true;
+      const unsigned int dim = _qrule_msm->get_dim();
+      const QuadratureType type = _qrule_msm->type();
+      delete _qrule_msm;
+
+      _const_qrule_msm = _qrule_msm = QBase::build(type, dim, order).release();
+      _fe_msm->attach_quadrature_rule(_qrule_msm);
```

Comment:
The mortar quadrature rule is defined when Assembly is initialized and before the MortarConstraint objects exist, so if a different one is specified in the MortarConstraint this will just swap a new rule in for the old one.